src/core/ad.jl

```diff
@@ -108,9 +108,11 @@ function gradient_logp(
     logp_old = getlogp(vi)
     function f(θ)
         new_vi = VarInfo(vi, sampler, θ)
-        model(new_vi, sampler, ctx)
+        new_vi = last(DynamicPPL.evaluate!!(model, new_vi, sampler, ctx))
         logp = getlogp(new_vi)
-        setlogp!(vi, ForwardDiff.value(logp))
+        # Don't need to capture the resulting `vi` since this is only
+        # needed if `vi` is mutable.
+        setlogp!!(vi, ForwardDiff.value(logp))
```

Comment:
I guess ideally we would want a no-op here when `setlogp!!` returns a new object - if we just call `setlogp!!` with a SimpleVarInfo it would create a new VarInfo object even though it's never used. Maybe one could check if `vi === new_vi` and only call `setlogp!!` if it is `true` (assuming that in this case `setlogp!!` would not return a new object)?

But I guess this should be left for the SimpleVarInfo PR?

Reply:
Will this not be optimized away in the case where `vi` is immutable?
I sort of assumed so. But I guess we might as well give the compiler a helping hand just to be sure.